### PR TITLE
Add:registry directory extension

### PR DESCRIPTION
**Pull request by hxmhlt**
**What this PR does**:
registry directory extension, the 
**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by hxmhlt** [comment]
U can run the example of `overriding registry_directory` from https://github.com/hxmhlt/dubbo-samples/tree/override_registry_directory/golang/override_registry_directory/dubbo/go-client

File: registry/directory/directory.go [hunk 1]
```diff
@@ -244,8 +233,8 @@ func (dir *registryDirectory) cacheInvoker(url *common.URL) protocol.Invoker {
 	return nil
 }
 
-//select the protocol invokers from the directory
-func (dir *registryDirectory) List(invocation protocol.Invocation) []protocol.Invoker {
+// list :select the protocol invokers from the directory
```
Comment (zouyx) [inline registry/directory/directory.go:236]:
```suggestion
// List selected protocol invokers from the directory
```

File: registry/directory/directory.go [hunk 2]
```diff
@@ -61,48 +57,41 @@ type registryDirectory struct {
...
 }
 
-func (dir *registryDirectory) Notify(event *registry.ServiceEvent) {
+func (dir *RegistryDirectory) Notify(event *registry.ServiceEvent) {
 	go dir.update(event)
 }
 
-//subscribe service from registry, and update the cacheServices
-func (dir *registryDirectory) update(res *registry.ServiceEvent) {
+// update: subscribe service from registry, and update the cacheServices
```
Comment (zouyx) [inline registry/directory/directory.go:93]:
```suggestion
// update the cacheServices and subscribe service from registry
```

File: registry/directory/directory.go [hunk 3]
```diff
@@ -198,8 +187,8 @@ func (dir *registryDirectory) toGroupInvokers() []protocol.Invoker {
 	return groupInvokersList
 }
 
-// uncacheInvoker return abandoned Invoker,if no Invoker to be abandoned,return nil
-func (dir *registryDirectory) uncacheInvoker(url *common.URL) protocol.Invoker {
+// uncacheInvoker: return abandoned Invoker,if no Invoker to be abandoned,return nil
```
Comment (zouyx) [inline registry/directory/directory.go:190]:
```suggestion
// uncacheInvoker will return abandoned Invoker,if no Invoker to be abandoned,return nil
```

File: registry/directory/directory.go [hunk 4]
```diff
@@ -208,8 +197,8 @@ func (dir *registryDirectory) uncacheInvoker(url *common.URL) protocol.Invoker {
 	return nil
 }
 
-// cacheInvoker return abandoned Invoker,if no Invoker to be abandoned,return nil
-func (dir *registryDirectory) cacheInvoker(url *common.URL) protocol.Invoker {
+// cacheInvoker: return abandoned Invoker,if no Invoker to be abandoned,return nil
```
Comment (zouyx) [inline registry/directory/directory.go:200]:
```suggestion
// cacheInvoker will return abandoned Invoker,if no Invoker to be abandoned,return nil
```

File: common/extension/registry_directory.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+package extension
+
+import (
+	"github.com/apache/dubbo-go/cluster"
+	"github.com/apache/dubbo-go/common"
+	"github.com/apache/dubbo-go/registry"
+)
+
+var defaultRegistry func(url *common.URL, registry registry.Registry) (cluster.Directory, error)
```
Comment (pantianying) [inline common/extension/registry_directory.go:26]: 这个方法类型的出参为啥用cluster.Directory，而不是registryDirectory

Comment (hxmhlt) [reply common/extension/registry_directory.go:26]: 用的是接口类，如果返回registryDirectory，那就无扩展性了。

**Pull request merged**